File: src/protocol.c
#include "protocol.h"
#include "crypto.h"

#include <stdlib.h>
#include <string.h>

/* ── Packet ──────────────────────────────────────────────────────── */

void packet_init(Packet *p)
{
    memset(p, 0, sizeof(*p));
    p->src     = 0x21;
    p->dst     = 0x01;
    p->dsrc    = 1;
    p->ddst    = 1;
    p->version = 3;
}

size_t packet_to_bytes(const Packet *p, uint8_t *buf, size_t buf_size)
{
    /* header (4) + hcrc (1) + product (1) + seq (4) + pad (2) +
       src/dst (2) + dsrc/ddst (2) + cmd (2) + payload + crc16 (2) */
    size_t need = 18 + (p->version >= 3 ? 2 : 0) + p->payload_len + 2;
    if (buf_size < need)
        return 0;

    size_t pos = 0;

    /* Header: prefix, version, payload length */
    buf[pos++] = PACKET_PREFIX;
    buf[pos++] = p->version;
    buf[pos++] = (uint8_t)(p->payload_len & 0xFF);
    buf[pos++] = (uint8_t)((p->payload_len >> 8) & 0xFF);

    /* Header CRC8 over first 4 bytes */
    buf[pos] = crc8_ccitt(buf, 4);
    pos++;

    /* Product byte */
    buf[pos++] = (p->product_id >= 0) ? 0x0D : 0x0C;

    /* Sequence */
    memcpy(buf + pos, p->seq, 4);
    pos += 4;

    /* Static padding */
    buf[pos++] = 0x00;
    buf[pos++] = 0x00;

    /* Source / destination */
    buf[pos++] = p->src;
    buf[pos++] = p->dst;

    /* V3+ dsrc/ddst */
    if (p->version >= 3) {
        buf[pos++] = p->dsrc;
        buf[pos++] = p->ddst;
    }

    /* Command set / id */
    buf[pos++] = p->cmd_set;
    buf[pos++] = p->cmd_id;

    /* Payload */
    if (p->payload_len > 0 && p->payload)
        memcpy(buf + pos, p->payload, p->payload_len);
    pos += p->payload_len;

    /* CRC16 over everything so far */
    uint16_t crc = crc16_arc(buf, pos);
    buf[pos++] = (uint8_t)(crc & 0xFF);
    buf[pos++] = (uint8_t)((crc >> 8) & 0xFF);

    return pos;
}

bool packet_from_bytes(const uint8_t *data, size_t len, Packet *p)
{
    if (len < 4 || data[0] != PACKET_PREFIX)
        return false;

    uint8_t version = data[1];
    size_t min_len = (version == 2) ? 18 : 20;
    if (len < min_len)
        return false;

    uint16_t payload_length = (uint16_t)data[2] | ((uint16_t)data[3] << 8);

    /* Verify CRC16 */
    if (version == 2 || version == 3 || version == 4) {
        uint16_t stored_crc = (uint16_t)data[len - 2] |
                              ((uint16_t)data[len - 1] << 8);
        if (crc16_arc(data, len - 2) != stored_crc)
            return false;
    }

    /* Verify header CRC8 */
    if (crc8_ccitt(data, 4) != data[4])
        return false;

    memset(p, 0, sizeof(*p));
    p->version = version;
    memcpy(p->seq, data + 6, 4);
    p->src = data[12];
    p->dst = data[13];

    size_t payload_start;
    if (version == 2) {
        p->dsrc = 0;
        p->ddst = 0;
        p->cmd_set = data[14];
        p->cmd_id  = data[15];
        payload_start = 16;
    } else {
        p->dsrc    = data[14];
        p->ddst    = data[15];
        p->cmd_set = data[16];
        p->cmd_id  = data[17];
        payload_start = 18;
    }

    p->payload_len = payload_length;
    p->payload = NULL;

    if (payload_length > 0) {
        p->payload = malloc(payload_length);
        if (!p->payload)
            return false;
        memcpy(p->payload, data + payload_start, payload_length);
    }

    return true;
}

/* ── EncPacket ───────────────────────────────────────────────────── */

uint8_t *enc_packet_build(const uint8_t *inner, size_t inner_len,
                          uint8_t frame_type,
                          const uint8_t *enc_key, const uint8_t *iv,
                          size_t *out_len)
{
    const uint8_t *payload = inner;
    size_t payload_len = inner_len;
    uint8_t *encrypted = NULL;

    if (enc_key && iv) {
        encrypted = aes_encrypt(inner, inner_len, enc_key, iv, &payload_len);
        if (!encrypted)
            return NULL;
        payload = encrypted;
    }

    /* header(6) + payload + crc16(2) */
    size_t total = 6 + payload_len + 2;
    uint8_t *buf = malloc(total);
    if (!buf) {
        free(encrypted);
        return NULL;
    }

    size_t pos = 0;
    buf[pos++] = ENC_PACKET_PREFIX_0;
    buf[pos++] = ENC_PACKET_PREFIX_1;
    buf[pos++] = (uint8_t)(frame_type << 4);
    buf[pos++] = 0x01; /* unknown byte */
    uint16_t plen = (uint16_t)(payload_len + 2); /* +2 for trailing CRC */
    buf[pos++] = (uint8_t)(plen & 0xFF);
    buf[pos++] = (uint8_t)((plen >> 8) & 0xFF);

    memcpy(buf + pos, payload, payload_len);
    pos += payload_len;

    uint16_t crc = crc16_arc(buf, pos);
    buf[pos++] = (uint8_t)(crc & 0xFF);
    buf[pos++] = (uint8_t)((crc >> 8) & 0xFF);

    free(encrypted);
    *out_len = pos;
    return buf;
}

/* ── Protobuf decoder ────────────────────────────────────────────── */

static bool read_varint(const uint8_t *data, size_t len, size_t *pos,
                        uint64_t *val)
{
    *val = 0;
    int shift = 0;
    while (*pos < len) {
        uint8_t b = data[(*pos)++];
        *val |= (uint64_t)(b & 0x7F) << shift;
        if (!(b & 0x80))
            return true;
        shift += 7;
        if (shift >= 64)
            return false;
    }
    return false;
}

size_t protobuf_decode(const uint8_t *data, size_t len,
                       PBField *fields, size_t max_fields)
{
    size_t pos = 0, count = 0;

    while (pos < len && count < max_fields) {
        uint64_t tag;
        if (!read_varint(data, len, &pos, &tag))
            break;

        uint32_t field_num = (uint32_t)(tag >> 3);
        uint8_t  wire_type = (uint8_t)(tag & 0x07);

        if (wire_type == 0) { /* varint */
            uint64_t val;
            if (!read_varint(data, len, &pos, &val))
                break;
            fields[count].field_num = field_num;
            fields[count].wire_type = 0;
            fields[count].value.u64 = val;
            count++;
        } else if (wire_type == 5) { /* fixed32 / float */
            if (pos + 4 > len) break;
            float fval;
            memcpy(&fval, data + pos, 4);
            pos += 4;
            fields[count].field_num = field_num;
            fields[count].wire_type = 5;
            fields[count].value.f32 = fval;
            count++;
        } else if (wire_type == 1) { /* fixed64 / double */
            if (pos + 8 > len) break;
            double dval;
            memcpy(&dval, data + pos, 8);
            pos += 8;
            fields[count].field_num = field_num;
            fields[count].wire_type = 1;
            fields[count].value.f64 = dval;
            count++;
        } else if (wire_type == 2) { /* length-delimited – skip */
            uint64_t length;
            if (!read_varint(data, len, &pos, &length))
                break;
            pos += (size_t)length;
        } else {
            break;
        }
    }

    return count;
}

/* ── Status parser ───────────────────────────────────────────────── */

/* Find a protobuf field by number. Returns NULL if not found. */
static const PBField *find_field(const PBField *fields, size_t count,
                                 uint32_t num)
{
    for (size_t i = 0; i < count; i++) {
        if (fields[i].field_num == num)
            return &fields[i];
    }
    return NULL;
}

static uint64_t field_u64(const PBField *fields, size_t count, uint32_t num)
{
    const PBField *f = find_field(fields, count, num);
    return f ? f->value.u64 : 0;
}

static float field_float(const PBField *fields, size_t count, uint32_t num)
{
    const PBField *f = find_field(fields, count, num);
    if (!f) return 0.0f;
    if (f->wire_type == 5)
        return f->value.f32;
    return (float)f->value.u64;
}

bool parse_river3_status(const uint8_t *data, size_t len, River3Status *out)
{
    PBField fields[128];
    size_t count = protobuf_decode(data, len, fields, 128);
    if (count == 0)
        return false;

    memset(out, 0, sizeof(*out));

    out->ac_input_voltage = (float)field_u64(fields, count, 227);
    out->ac_input_power   = field_float(fields, count, 3);
    out->ac_output_power  = field_float(fields, count, 9);
    out->ac_plugged_in    = field_u64(fields, count, 227) > 0;

    /* battery_level: prefer field 262, fall back to 4 */
    uint64_t batt = field_u64(fields, count, 262);
    if (batt == 0)
        batt = field_u64(fields, count, 4);
    out->battery_level = (int)batt;

    out->battery_temp    = (float)field_u64(fields, count, 258);
    out->dc_input_power  = field_float(fields, count, 11);
    out->usb_output_power = field_float(fields, count, 12);

    return true;
}

File: src/protocol.ts
import { crc8Ccitt, crc16Arc, aesEncrypt } from "./crypto.js";

const PACKET_PREFIX = 0xaa;
const ENC_PACKET_PREFIX_0 = 0x5a;
const ENC_PACKET_PREFIX_1 = 0x5a;

export const FRAME_TYPE_COMMAND = 0x00;
export const FRAME_TYPE_PROTOCOL = 0x01;

export interface Packet {
  src: number;
  dst: number;
  cmdSet: number;
  cmdId: number;
  dsrc: number;
  ddst: number;
  version: number;
  seq: Uint8Array;
  productId: number;
  payload: Uint8Array;
}

export function packetInit(): Packet {
  return {
    src: 0x21,
    dst: 0x01,
    dsrc: 1,
    ddst: 1,
    version: 3,
    seq: new Uint8Array(4),
    productId: 0,
    payload: new Uint8Array(0),
    cmdSet: 0,
    cmdId: 0,
  };
}

export function packetToBytes(p: Packet): Uint8Array {
  const hasExtraAddr = p.version >= 3;
  const need = 18 + (hasExtraAddr ? 2 : 0) + p.payload.length + 2;
  const buf = new Uint8Array(need);
  let pos = 0;

  buf[pos++] = PACKET_PREFIX;
  buf[pos++] = p.version;
  buf[pos++] = p.payload.length & 0xff;
  buf[pos++] = (p.payload.length >> 8) & 0xff;

  buf[pos] = crc8Ccitt(buf.subarray(0, 4));
  pos++;

  buf[pos++] = p.productId >= 0 ? 0x0d : 0x0c;

  buf.set(p.seq, pos);
  pos += 4;

  buf[pos++] = 0x00;
  buf[pos++] = 0x00;

  buf[pos++] = p.src;
  buf[pos++] = p.dst;

  if (hasExtraAddr) {
    buf[pos++] = p.dsrc;
    buf[pos++] = p.ddst;
  }

  buf[pos++] = p.cmdSet;
  buf[pos++] = p.cmdId;

  if (p.payload.length > 0) {
    buf.set(p.payload, pos);
  }
  pos += p.payload.length;

  const crc = crc16Arc(buf.subarray(0, pos));
  buf[pos++] = crc & 0xff;
  buf[pos++] = (crc >> 8) & 0xff;

  return buf.subarray(0, pos);
}

export function packetFromBytes(data: Uint8Array): Packet | null {
  if (data.length < 4 || data[0] !== PACKET_PREFIX) return null;

  const version = data[1];
  const minLen = version === 2 ? 18 : 20;
  if (data.length < minLen) return null;

  const payloadLength = data[2] | (data[3] << 8);

  if (version === 2 || version === 3 || version === 4) {
    const storedCrc = data[data.length - 2] | (data[data.length - 1] << 8);
    if (crc16Arc(data.subarray(0, data.length - 2)) !== storedCrc) return null;
  }

  if (crc8Ccitt(data.subarray(0, 4)) !== data[4]) return null;

  const seq = new Uint8Array(data.subarray(6, 10));
  const src = data[12];
  const dst = data[13];

  let dsrc: number, ddst: number, cmdSet: number, cmdId: number;
  let payloadStart: number;

  if (version === 2) {
    dsrc = 0;
    ddst = 0;
    cmdSet = data[14];
    cmdId = data[15];
    payloadStart = 16;
  } else {
    dsrc = data[14];
    ddst = data[15];
    cmdSet = data[16];
    cmdId = data[17];
    payloadStart = 18;
  }

  const payload = new Uint8Array(
    data.subarray(payloadStart, payloadStart + payloadLength),
  );

  return {
    version,
    seq,
    src,
    dst,
    dsrc,
    ddst,
    cmdSet,
    cmdId,
    productId: 0,
    payload,
  };
}

export function encPacketBuild(
  inner: Uint8Array,
  frameType: number,
  encKey?: Uint8Array,
  iv?: Uint8Array,
): Uint8Array {
  let payload = inner;
  if (encKey && iv) {
    payload = aesEncrypt(inner, encKey, iv);
  }

  const total = 6 + payload.length + 2;
  const buf = new Uint8Array(total);
  let pos = 0;

  buf[pos++] = ENC_PACKET_PREFIX_0;
  buf[pos++] = ENC_PACKET_PREFIX_1;
  buf[pos++] = frameType << 4;
  buf[pos++] = 0x01;
  const plen = payload.length + 2;
  buf[pos++] = plen & 0xff;
  buf[pos++] = (plen >> 8) & 0xff;

  buf.set(payload, pos);
  pos += payload.length;

  const crc = crc16Arc(buf.subarray(0, pos));
  buf[pos++] = crc & 0xff;
  buf[pos++] = (crc >> 8) & 0xff;

  return buf.subarray(0, pos);
}

interface PBField {
  fieldNum: number;
  wireType: number;
  value: number | bigint;
}

function readVarint(
  data: Uint8Array,
  pos: { v: number },
): bigint | null {
  let val = 0n;
  let shift = 0;
  while (pos.v < data.length) {
    const b = data[pos.v++];
    val |= BigInt(b & 0x7f) << BigInt(shift);
    if (!(b & 0x80)) return val;
    shift += 7;
    if (shift >= 64) return null;
  }
  return null;
}

function protobufDecode(data: Uint8Array): PBField[] {
  const fields: PBField[] = [];
  const pos = { v: 0 };

  while (pos.v < data.length) {
    const tag = readVarint(data, pos);
    if (tag === null) break;

    const fieldNum = Number(tag >> 3n);
    const wireType = Number(tag & 7n);

    if (wireType === 0) {
      const val = readVarint(data, pos);
      if (val === null) break;
      fields.push({ fieldNum, wireType, value: val });
    } else if (wireType === 5) {
      if (pos.v + 4 > data.length) break;
      const dv = new DataView(data.buffer, data.byteOffset + pos.v, 4);
      fields.push({ fieldNum, wireType, value: dv.getFloat32(0, true) });
      pos.v += 4;
    } else if (wireType === 1) {
      if (pos.v + 8 > data.length) break;
      const dv = new DataView(data.buffer, data.byteOffset + pos.v, 8);
      fields.push({ fieldNum, wireType, value: dv.getFloat64(0, true) });
      pos.v += 8;
    } else if (wireType === 2) {
      const length = readVarint(data, pos);
      if (length === null) break;
      pos.v += Number(length);
    } else {
      break;
    }
  }

  return fields;
}

export interface River3Status {
  batteryLevel: number;
  batteryTemp: number;
  acInputPower: number;
  acInputVoltage: number;
  acPluggedIn: boolean;
  acOutputPower: number;
  dcInputPower: number;
  usbOutputPower: number;
}

function findField(fields: PBField[], num: number): PBField | undefined {
  let last: PBField | undefined;
  for (const f of fields) {
    if (f.fieldNum === num) last = f;
  }
  return last;
}

function fieldFloat(fields: PBField[], num: number): number {
  const f = findField(fields, num);
  if (!f) return 0;
  return Number(f.value);
}

export function parseRiver3Status(data: Uint8Array): River3Status | null {
  const fields = protobufDecode(data);
  if (fields.length === 0) return null;

  const acInputVoltage = fieldFloat(fields, 227);
  const acInputPower = fieldFloat(fields, 54);
  const acOutputPower = fieldFloat(fields, 4);

  let batteryLevel = fieldFloat(fields, 262);
  if (batteryLevel === 0) batteryLevel = fieldFloat(fields, 4);

  return {
    acInputVoltage,
    acInputPower,
    acOutputPower,
    acPluggedIn: acInputPower > 0,
    batteryLevel,
    batteryTemp: fieldFloat(fields, 258),
    dcInputPower: fieldFloat(fields, 11),
    usbOutputPower: fieldFloat(fields, 12),
  };
}

export function gridAvailable(s: River3Status): boolean {
  return s.acInputPower > 1;
}

export {
  PACKET_PREFIX,
  ENC_PACKET_PREFIX_0,
  ENC_PACKET_PREFIX_1,
};
